library/core/src/panic.rs

```diff
@@ -140,6 +140,36 @@ pub macro unreachable_2021 {
     ),
 }
 
+/// Invokes a closure, aborting if the closure unwinds.
+///
+/// When compiled with aborting panics, this function is effectively a no-op.
+/// With unwinding panics, an unwind results in another call into the panic
+/// hook followed by a process abort.
+///
+/// # Notes
+///
+/// Instead of using this function, code should attempt to support unwinding.
+/// Implementing [`Drop`] allows you to restore invariants uniformly in both
+/// return and unwind paths.
+///
+/// If an unwind can lead to logical issues but not soundness issues, you
+/// should allow the unwind. Opting out of [`UnwindSafe`] indicates to your
+/// consumers that they need to consider correctness in the face of unwinds.
+///
+/// If an unwind would be unsound, then this function should be used in order
+/// to prevent unwinds. However, note that `extern "C" fn` will automatically
+/// convert unwinds to aborts, so using this function isn't necessary for FFI.
+#[unstable(feature = "abort_unwind", issue = "130338")]
+pub fn abort_unwind<F: FnOnce() -> R, R>(f: F) -> R {
+    // This attribute adds the "unwinding out of nounwind function" guard.
+    #[rustc_nounwind]
+    fn abort_unwind_inner<F: FnOnce() -> R, R>(f: F) -> R {
+        f()
+    }
+
+    abort_unwind_inner(f)
+}
```

Comment:
Is this 100% interchangeable with the following&mdash;? I don't immediately understand the purpose of `abort_unwind_inner`. Or does moving `rustc_nounwind` to the public function cause some observable consequence like the function can no longer be coerced to an extern "Rust" function pointer.

```suggestion
#[rustc_nounwind]
pub fn abort_unwind<F: FnOnce() -> R, R>(f: F) -> R {
    f()
}
```

Reply:
`#[rustc_nounwind]` doesn't impact function pointer usage (like ABI does), nor does it seem to show up in rustdoc (like I thought it might). I conservatively used the inner function to avoid annotating the `pub` item with an internal rustc implementation detail (it's only on non-exported items currently), but since it does seem to be a private detail, avoiding the extra layer is better.

This can always be changed later if we turn out to have been wrong about this.